emmarin/cl/cl/src/cl/note.rs

```diff
@@ -0,0 +1,171 @@
+use rand::RngCore;
+use serde::{Deserialize, Serialize};
+use sha2::{Digest, Sha256};
+
+use crate::cl::{balance::Unit, nullifier::NullifierCommitment};
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+pub struct Constraint(pub [u8; 32]);
+
+impl Constraint {
+    pub fn from_vk(constraint_vk: &[u8]) -> Self {
+        let mut hasher = Sha256::new();
+        hasher.update(b"NOMOS_CL_CONSTRAINT_COMMIT");
+        hasher.update(constraint_vk);
+        let constraint_cm: [u8; 32] = hasher.finalize().into();
+
+        Self(constraint_cm)
+    }
+}
+
+pub fn derive_unit(unit: &str) -> Unit {
+    let mut hasher = Sha256::new();
+    hasher.update(b"NOMOS_CL_UNIT");
+    hasher.update(unit.as_bytes());
+    let unit: Unit = hasher.finalize().into();
+    unit
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+pub struct NoteCommitment(pub [u8; 32]);
+
+impl NoteCommitment {
+    pub fn as_bytes(&self) -> &[u8; 32] {
+        &self.0
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy, Serialize, Deserialize)]
+pub struct NoteWitness {
+    pub value: u64,
+    pub unit: Unit,
+    pub constraint: Constraint,
+    pub state: [u8; 32],
+    pub nonce: Nonce,
+}
+
+impl NoteWitness {
+    pub fn new(
+        value: u64,
+        unit: Unit,
+        constraint: Constraint,
+        state: [u8; 32],
+        nonce: Nonce,
+    ) -> Self {
+        Self {
+            value,
+            unit,
+            constraint,
+            state,
+            nonce,
+        }
+    }
+
+    pub fn basic(value: u64, unit: Unit, rng: impl RngCore) -> Self {
+        let constraint = Constraint([0u8; 32]);
+        let nonce = Nonce::random(rng);
+        Self::new(value, unit, constraint, [0u8; 32], nonce)
+    }
+
+    pub fn stateless(value: u64, unit: Unit, constraint: Constraint, rng: impl RngCore) -> Self {
+        Self::new(value, unit, constraint, [0u8; 32], Nonce::random(rng))
+    }
+
+    pub fn commit(&self, tag: &dyn AsRef<[u8]>, nf_pk: NullifierCommitment) -> NoteCommitment {
+        let mut hasher = Sha256::new();
+        hasher.update(tag.as_ref());
```

Comment:
We still need a DST here e.g. `"NOMOS_CL_NOTE_CM"` and it should be constant, if there is something variable at runtime, it should be part of the hash payload